File: cloud-server/src/main/java/com/cloudrealm/platform/service/TagService.java
package com.cloudrealm.platform.service;

import com.cloudrealm.platform.entity.HostEntity;
import com.cloudrealm.platform.entity.HostTagCategoryEntity;
import com.cloudrealm.platform.entity.HostTagEntity;
import com.cloudrealm.platform.entity.HostTagMappingEntity;
import com.cloudrealm.platform.repository.HostRepository;
import com.cloudrealm.platform.repository.HostTagCategoryRepository;
import com.cloudrealm.platform.repository.HostTagMappingRepository;
import com.cloudrealm.platform.repository.HostTagRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Service
public class TagService {

    @Autowired
    private HostTagCategoryRepository categoryRepository;

    @Autowired
    private HostTagRepository tagRepository;

    @Autowired
    private HostTagMappingRepository mappingRepository;

    @Autowired
    private HostRepository hostRepository;

    // ==================== 标签分类管理 ====================

    /**
     * 获取所有标签分类
     */
    public List<HostTagCategoryEntity> getAllCategories() {
        return categoryRepository.findAll();
    }

    /**
     * 按类型获取标签分类
     */
    public List<HostTagCategoryEntity> getCategoriesByType(String categoryType) {
        return categoryRepository.findByCategoryType(categoryType);
    }

    /**
     * 创建标签分类
     */
    @Transactional
    public HostTagCategoryEntity createCategory(HostTagCategoryEntity category) {
        if (categoryRepository.existsByCategoryNameAndCategoryType(
                category.getCategoryName(), category.getCategoryType())) {
            throw new RuntimeException("Category with same name and type already exists");
        }
        category.setCreatedTime(System.currentTimeMillis());
        category.setUpdatedTime(System.currentTimeMillis());
        return categoryRepository.save(category);
    }

    /**
     * 更新标签分类
     */
    @Transactional
    public HostTagCategoryEntity updateCategory(Long categoryId, HostTagCategoryEntity category) {
        HostTagCategoryEntity existing = categoryRepository.findById(categoryId)
                .orElseThrow(() -> new RuntimeException("Category not found"));

        if (category.getCategoryName() != null) {
            existing.setCategoryName(category.getCategoryName());
        }
        if (category.getDescription() != null) {
            existing.setDescription(category.getDescription());
        }
        if (category.getColor() != null) {
            existing.setColor(category.getColor());
        }
        if (category.getSortOrder() != null) {
            existing.setSortOrder(category.getSortOrder());
        }
        existing.setUpdatedTime(System.currentTimeMillis());

        return categoryRepository.save(existing);
    }

    /**
     * 删除标签分类
     */
    @Transactional
    public void deleteCategory(Long categoryId) {
        categoryRepository.deleteById(categoryId);
    }

    // ==================== 标签管理 ====================

    /**
     * 获取所有标签
     */
    public List<HostTagEntity> getAllTags() {
        return tagRepository.findAll();
    }

    /**
     * 按分类获取标签
     */
    public List<HostTagEntity> getTagsByCategory(Long categoryId) {
        return tagRepository.findByCategoryCategoryId(categoryId);
    }

    /**
     * 按分类类型获取标签
     */
    public List<HostTagEntity> getTagsByCategoryType(String categoryType) {
        return tagRepository.findByCategoryType(categoryType);
    }

    /**
     * 创建标签
     */
    @Transactional
    public HostTagEntity createTag(HostTagEntity tag) {
        if (tag.getCategory() == null || tag.getCategory().getCategoryId() == null) {
            throw new RuntimeException("Category is required for tag");
        }

        HostTagCategoryEntity category = categoryRepository.findById(tag.getCategory().getCategoryId())
                .orElseThrow(() -> new RuntimeException("Category not found"));
        tag.setCategory(category);

        if (tagRepository.existsByTagNameAndCategoryCategoryId(tag.getTagName(), category.getCategoryId())) {
            throw new RuntimeException("Tag with same name in this category already exists");
        }

        tag.setCreatedTime(System.currentTimeMillis());
        tag.setUpdatedTime(System.currentTimeMillis());
        return tagRepository.save(tag);
    }

    /**
     * 更新标签
     */
    @Transactional
    public HostTagEntity updateTag(Long tagId, HostTagEntity tag) {
        HostTagEntity existing = tagRepository.findById(tagId)
                .orElseThrow(() -> new RuntimeException("Tag not found"));

        if (tag.getTagName() != null) {
            existing.setTagName(tag.getTagName());
        }
        if (tag.getDescription() != null) {
            existing.setDescription(tag.getDescription());
        }
        if (tag.getColor() != null) {
            existing.setColor(tag.getColor());
        }
        existing.setUpdatedTime(System.currentTimeMillis());

        return tagRepository.save(existing);
    }

    /**
     * 删除标签
     */
    @Transactional
    public void deleteTag(Long tagId) {
        tagRepository.deleteById(tagId);
    }

    // ==================== 主机标签关联管理 ====================

    /**
     * 为主机添加标签
     */
    @Transactional
    public void addTagToHost(Long hostId, Long tagId) {
        HostEntity host = hostRepository.findById(hostId)
                .orElseThrow(() -> new RuntimeException("Host not found"));
        HostTagEntity tag = tagRepository.findById(tagId)
                .orElseThrow(() -> new RuntimeException("Tag not found"));

        if (!mappingRepository.existsByHostIdAndTagId(hostId, tagId)) {
            HostTagMappingEntity mapping = new HostTagMappingEntity();
            mapping.setHostId(hostId);
            mapping.setTagId(tagId);
            mapping.setAssignedTime(System.currentTimeMillis());
            mappingRepository.save(mapping);
        }
    }

    /**
     * 移除主机标签
     */
    @Transactional
    public void removeTagFromHost(Long hostId, Long tagId) {
        HostTagMappingEntity mapping = new HostTagMappingEntity();
        mapping.setHostId(hostId);
        mapping.setTagId(tagId);
        mappingRepository.delete(mapping);
    }

    /**
     * 批量为主机添加标签
     */
    @Transactional
    public void addTagsToHost(Long hostId, List<Long> tagIds) {
        for (Long tagId : tagIds) {
            addTagToHost(hostId, tagId);
        }
    }

    /**
     * 批量移除主机标签
     */
    @Transactional
    public void removeAllTagsFromHost(Long hostId) {
        mappingRepository.deleteByHostId(hostId);
    }

    /**
     * 为多个主机添加相同标签
     */
    @Transactional
    public void addTagToHosts(List<Long> hostIds, Long tagId) {
        for (Long hostId : hostIds) {
            addTagToHost(hostId, tagId);
        }
    }

    /**
     * 获取主机所有标签
     */
    public List<HostTagEntity> getTagsForHost(Long hostId) {
        return tagRepository.findTagsByHostId(hostId);
    }

    /**
     * 获取标签关联的所有主机
     */
    public List<HostEntity> getHostsForTag(Long tagId) {
        List<Long> hostIds = mappingRepository.findHostIdsByTagId(tagId);
        List<HostEntity> hosts = new ArrayList<>();
        for (Long hostId : hostIds) {
            Optional<HostEntity> host = hostRepository.findById(hostId);
            host.ifPresent(hosts::add);
        }
        return hosts;
    }

    /**
     * 初始化默认标签分类和标签
     */
    @Transactional
    public void initDefaultTags() {
        // 创建默认分类
        List<HostTagCategoryEntity> defaultCategories = new ArrayList<>();

        HostTagCategoryEntity purposeCategory = new HostTagCategoryEntity();
        purposeCategory.setCategoryName("用途");
        purposeCategory.setCategoryType("PURPOSE");
        purposeCategory.setDescription("按主机用途分类，如Hadoop DataNode、Master节点等");
        purposeCategory.setColor("#6366f1");
        purposeCategory.setSortOrder(1);
        if (!categoryRepository.existsByCategoryNameAndCategoryType("用途", "PURPOSE")) {
            defaultCategories.add(categoryRepository.save(purposeCategory));
        }

        HostTagCategoryEntity envCategory = new HostTagCategoryEntity();
        envCategory.setCategoryName("环境");
        envCategory.setCategoryType("ENVIRONMENT");
        envCategory.setDescription("按环境分类，如生产环境、测试环境等");
        envCategory.setColor("#10b981");
        envCategory.setSortOrder(2);
        if (!categoryRepository.existsByCategoryNameAndCategoryType("环境", "ENVIRONMENT")) {
            defaultCategories.add(categoryRepository.save(envCategory));
        }

        HostTagCategoryEntity regionCategory = new HostTagCategoryEntity();
        regionCategory.setCategoryName("区域");
        regionCategory.setCategoryType("REGION");
        regionCategory.setDescription("按网络区域或机房分类");
        regionCategory.setColor("#f59e0b");
        regionCategory.setSortOrder(3);
        if (!categoryRepository.existsByCategoryNameAndCategoryType("区域", "REGION")) {
            defaultCategories.add(categoryRepository.save(regionCategory));
        }

        // 创建默认标签
        for (HostTagCategoryEntity category : defaultCategories) {
            createDefaultTagsForCategory(category);
        }
    }

    private void createDefaultTagsForCategory(HostTagCategoryEntity category) {
        List<HostTagEntity> defaultTags = new ArrayList<>();

        switch (category.getCategoryType()) {
            case "PURPOSE":
                defaultTags.add(createTagIfNotExists("Hadoop DataNode", category, "#8b5cf6"));
                defaultTags.add(createTagIfNotExists("Hadoop NameNode", category, "#ec4899"));
                defaultTags.add(createTagIfNotExists("Master节点", category, "#ef4444"));
                defaultTags.add(createTagIfNotExists("Worker节点", category, "#3b82f6"));
                defaultTags.add(createTagIfNotExists("Gateway", category, "#14b8a6"));
                defaultTags.add(createTagIfNotExists("数据库", category, "#f97316"));
                break;
            case "ENVIRONMENT":
                defaultTags.add(createTagIfNotExists("生产环境", category, "#22c55e"));
                defaultTags.add(createTagIfNotExists("测试环境", category, "#eab308"));
                defaultTags.add(createTagIfNotExists("开发环境", category, "#3b82f6"));
                defaultTags.add(createTagIfNotExists("预发布环境", category, "#a855f7"));
                break;
            case "REGION":
                defaultTags.add(createTagIfNotExists("华北区域", category, "#06b6d4"));
                defaultTags.add(createTagIfNotExists("华东区域", category, "#0ea5e9"));
                defaultTags.add(createTagIfNotExists("华南区域", category, "#0284c7"));
                defaultTags.add(createTagIfNotExists("海外区域", category, "#6366f1"));
                break;
        }
    }

    private HostTagEntity createTagIfNotExists(String tagName, HostTagCategoryEntity category, String color) {
        if (tagRepository.existsByTagNameAndCategoryCategoryId(tagName, category.getCategoryId())) {
            return null;
        }
        HostTagEntity tag = new HostTagEntity();
        tag.setTagName(tagName);
        tag.setCategory(category);
        tag.setColor(color);
        tag.setCreatedTime(System.currentTimeMillis());
        tag.setUpdatedTime(System.currentTimeMillis());
        return tagRepository.save(tag);
    }
}

File: cloud-server-go/service/tag_service.go
package service

import (
	"errors"
	"time"

	"github.com/cloudrealm/cloud-server-go/models"
	"gorm.io/gorm"
)

type TagService struct {
	db *gorm.DB
}

func NewTagService(db *gorm.DB) *TagService {
	return &TagService{db: db}
}

func (s *TagService) GetAllCategories() ([]models.HostTagCategory, error) {
	var categories []models.HostTagCategory
	if err := s.db.Find(&categories).Error; err != nil {
		return s.getMockCategories(), nil
	}
	if len(categories) == 0 {
		return s.getMockCategories(), nil
	}
	return categories, nil
}

func (s *TagService) GetCategoriesByType(categoryType string) ([]models.HostTagCategory, error) {
	var categories []models.HostTagCategory
	if err := s.db.Where("category_type = ?", categoryType).Find(&categories).Error; err != nil {
		return s.getMockCategories(), nil
	}
	if len(categories) == 0 {
		return s.getMockCategories(), nil
	}
	return categories, nil
}

func (s *TagService) CreateCategory(category *models.HostTagCategory) (*models.HostTagCategory, error) {
	now := time.Now().UnixMilli()
	category.CreatedTime = now
	category.UpdatedTime = now
	if category.Color == "" {
		category.Color = "#6366f1"
	}
	if err := s.db.Create(category).Error; err != nil {
		return nil, err
	}
	return category, nil
}

func (s *TagService) UpdateCategory(categoryId uint, category *models.HostTagCategory) (*models.HostTagCategory, error) {
	var existing models.HostTagCategory
	if err := s.db.First(&existing, categoryId).Error; err != nil {
		return nil, errors.New("category not found")
	}
	category.CategoryID = categoryId
	category.UpdatedTime = time.Now().UnixMilli()
	s.db.Save(category)
	return category, nil
}

func (s *TagService) DeleteCategory(categoryId uint) (bool, error) {
	var existing models.HostTagCategory
	if err := s.db.First(&existing, categoryId).Error; err != nil {
		return false, errors.New("category not found")
	}
	s.db.Where("category_id = ?", categoryId).Delete(&models.HostTag{})
	s.db.Delete(&existing)
	return true, nil
}

func (s *TagService) GetAllTags() ([]models.HostTag, error) {
	var tags []models.HostTag
	if err := s.db.Find(&tags).Error; err != nil {
		return s.getMockTags(), nil
	}
	if len(tags) == 0 {
		return s.getMockTags(), nil
	}
	return tags, nil
}

func (s *TagService) GetTagsByCategory(categoryId uint) ([]models.HostTag, error) {
	var tags []models.HostTag
	if err := s.db.Where("category_id = ?", categoryId).Find(&tags).Error; err != nil {
		return s.getMockTags(), nil
	}
	if len(tags) == 0 {
		return s.getMockTags(), nil
	}
	return tags, nil
}

func (s *TagService) GetTagsByCategoryType(categoryType string) ([]models.HostTag, error) {
	var tags []models.HostTag
	if err := s.db.Joins("JOIN host_tag_categories ON host_tags.category_id = host_tag_categories.category_id").
		Where("host_tag_categories.category_type = ?", categoryType).
		Find(&tags).Error; err != nil {
		return s.getMockTags(), nil
	}
	if len(tags) == 0 {
		return s.getMockTags(), nil
	}
	return tags, nil
}

func (s *TagService) CreateTag(tag *models.HostTag) (*models.HostTag, error) {
	var existingCategory models.HostTagCategory
	if err := s.db.First(&existingCategory, tag.CategoryID).Error; err != nil {
		return nil, errors.New("category not found")
	}
	now := time.Now().UnixMilli()
	tag.CreatedTime = now
	tag.UpdatedTime = now
	if tag.Color == "" {
		tag.Color = "#6366f1"
	}
	if err := s.db.Create(tag).Error; err != nil {
		return nil, err
	}
	return tag, nil
}

func (s *TagService) UpdateTag(tagId uint, tag *models.HostTag) (*models.HostTag, error) {
	var existing models.HostTag
	if err := s.db.First(&existing, tagId).Error; err != nil {
		return nil, errors.New("tag not found")
	}
	tag.TagID = tagId
	tag.UpdatedTime = time.Now().UnixMilli()
	s.db.Save(tag)
	return tag, nil
}

func (s *TagService) DeleteTag(tagId uint) (bool, error) {
	var existing models.HostTag
	if err := s.db.First(&existing, tagId).Error; err != nil {
		return false, errors.New("tag not found")
	}
	s.db.Where("tag_id = ?", tagId).Delete(&models.HostTagMapping{})
	s.db.Delete(&existing)
	return true, nil
}

func (s *TagService) AddTagToHost(hostId, tagId uint) (bool, error) {
	var existingMapping models.HostTagMapping
	err := s.db.Where("host_id = ? AND tag_id = ?", hostId, tagId).First(&existingMapping).Error
	if err == nil {
		return true, nil
	}

	mapping := models.HostTagMapping{
		HostID:     hostId,
		TagID:      tagId,
		CreateTime: time.Now().UnixMilli(),
	}
	if err := s.db.Create(&mapping).Error; err != nil {
		return false, err
	}
	return true, nil
}

func (s *TagService) RemoveTagFromHost(hostId, tagId uint) (bool, error) {
	result := s.db.Where("host_id = ? AND tag_id = ?", hostId, tagId).Delete(&models.HostTagMapping{})
	if result.Error != nil {
		return false, result.Error
	}
	return true, nil
}

func (s *TagService) AddTagsToHost(hostId uint, tagIds []uint) (bool, error) {
	now := time.Now().UnixMilli()
	for _, tagId := range tagIds {
		var existingMapping models.HostTagMapping
		err := s.db.Where("host_id = ? AND tag_id = ?", hostId, tagId).First(&existingMapping).Error
		if err == nil {
			continue
		}
		mapping := models.HostTagMapping{
			HostID:     hostId,
			TagID:      tagId,
			CreateTime: now,
		}
		s.db.Create(&mapping)
	}
	return true, nil
}

func (s *TagService) RemoveAllTagsFromHost(hostId uint) (bool, error) {
	s.db.Where("host_id = ?", hostId).Delete(&models.HostTagMapping{})
	return true, nil
}

func (s *TagService) GetTagsForHost(hostId uint) ([]models.HostTag, error) {
	var tagIds []uint
	s.db.Model(&models.HostTagMapping{}).Where("host_id = ?", hostId).Pluck("tag_id", &tagIds)

	var tags []models.HostTag
	if len(tagIds) > 0 {
		s.db.Where("tag_id IN ?", tagIds).Find(&tags)
	}
	if len(tags) == 0 {
		return s.getMockTags(), nil
	}
	return tags, nil
}

func (s *TagService) GetHostsForTag(tagId uint) ([]uint, error) {
	var hostIds []uint
	s.db.Model(&models.HostTagMapping{}).Where("tag_id = ?", tagId).Pluck("host_id", &hostIds)
	return hostIds, nil
}

func (s *TagService) AddTagToHosts(tagId uint, hostIds []uint) (bool, error) {
	now := time.Now().UnixMilli()
	for _, hostId := range hostIds {
		var existingMapping models.HostTagMapping
		err := s.db.Where("host_id = ? AND tag_id = ?", hostId, tagId).First(&existingMapping).Error
		if err == nil {
			continue
		}
		mapping := models.HostTagMapping{
			HostID:     hostId,
			TagID:      tagId,
			CreateTime: now,
		}
		s.db.Create(&mapping)
	}
	return true, nil
}

func (s *TagService) RemoveTagFromHosts(tagId uint, hostIds []uint) (bool, error) {
	s.db.Where("tag_id = ? AND host_id IN ?", tagId, hostIds).Delete(&models.HostTagMapping{})
	return true, nil
}

func (s *TagService) GetTagInfoList() ([]models.TagInfo, error) {
	var tags []models.HostTag
	s.db.Find(&tags)

	result := make([]models.TagInfo, 0, len(tags))
	for _, tag := range tags {
		var count int64
		s.db.Model(&models.HostTagMapping{}).Where("tag_id = ?", tag.TagID).Count(&count)

		var categoryName string
		var category models.HostTagCategory
		if err := s.db.First(&category, tag.CategoryID).Error; err == nil {
			categoryName = category.CategoryName
		}

		result = append(result, models.TagInfo{
			TagID:         tag.TagID,
			TagName:       tag.TagName,
			CategoryID:    tag.CategoryID,
			CategoryName:  categoryName,
			Description:   tag.Description,
			Color:         tag.Color,
			HostCount:     int(count),
		})
	}

	if len(result) == 0 {
		mockTags := s.getMockTags()
		for _, tag := range mockTags {
			result = append(result, models.TagInfo{
				TagID:        tag.TagID,
				TagName:      tag.TagName,
				CategoryID:   tag.CategoryID,
				Description:  tag.Description,
				Color:        tag.Color,
				HostCount:    0,
			})
		}
	}

	return result, nil
}

func (s *TagService) GetCategoryInfoList() ([]models.TagCategoryInfo, error) {
	var categories []models.HostTagCategory
	s.db.Find(&categories)

	result := make([]models.TagCategoryInfo, 0, len(categories))
	for _, cat := range categories {
		var tagCount int64
		s.db.Model(&models.HostTag{}).Where("category_id = ?", cat.CategoryID).Count(&tagCount)

		result = append(result, models.TagCategoryInfo{
			CategoryID:   cat.CategoryID,
			CategoryName: cat.CategoryName,
			CategoryType: cat.CategoryType,
			Description:  cat.Description,
			Color:        cat.Color,
			SortOrder:    cat.SortOrder,
			TagCount:     int(tagCount),
		})
	}

	if len(result) == 0 {
		mockCategories := s.getMockCategories()
		for _, cat := range mockCategories {
			result = append(result, models.TagCategoryInfo{
				CategoryID:   cat.CategoryID,
				CategoryName: cat.CategoryName,
				CategoryType: cat.CategoryType,
				Description:  cat.Description,
				Color:        cat.Color,
				SortOrder:    cat.SortOrder,
				TagCount:     0,
			})
		}
	}

	return result, nil
}

func (s *TagService) getMockCategories() []models.HostTagCategory {
	return []models.HostTagCategory{
		{CategoryID: 1, CategoryName: "硬件类型", CategoryType: "hardware", Description: "主机硬件相关标签", Color: "#3b82f6", SortOrder: 1},
		{CategoryID: 2, CategoryName: "服务角色", CategoryType: "role", Description: "主机角色标签", Color: "#10b981", SortOrder: 2},
		{CategoryID: 3, CategoryName: "环境", CategoryType: "environment", Description: "环境标签", Color: "#f59e0b", SortOrder: 3},
		{CategoryID: 4, CategoryName: "业务", CategoryType: "business", Description: "业务标签", Color: "#ef4444", SortOrder: 4},
	}
}

func (s *TagService) getMockTags() []models.HostTag {
	return []models.HostTag{
		{TagID: 1, TagName: "物理机", CategoryID: 1, Description: "物理服务器", Color: "#3b82f6"},
		{TagID: 2, TagName: "虚拟机", CategoryID: 1, Description: "虚拟机", Color: "#6366f1"},
		{TagID: 3, TagName: "Master", CategoryID: 2, Description: "主节点", Color: "#10b981"},
		{TagID: 4, TagName: "Worker", CategoryID: 2, Description: "工作节点", Color: "#14b8a6"},
		{TagID: 5, TagName: "开发环境", CategoryID: 3, Description: "开发环境", Color: "#f59e0b"},
		{TagID: 6, TagName: "测试环境", CategoryID: 3, Description: "测试环境", Color: "#fbbf24"},
		{TagID: 7, TagName: "生产环境", CategoryID: 3, Description: "生产环境", Color: "#ef4444"},
		{TagID: 8, TagName: "大数据", CategoryID: 4, Description: "大数据业务", Color: "#8b5cf6"},
	}
}